tasmanian@7.9 +cuda cuda_arch=90 fails build: InterfaceTPL/tsgCudaWrappers.cpp(301): identifier "CUSPARSE_MV_ALG_DEFAULT" is undefined

### Steps to reproduce the issue

`tasmanian@7.9 +cuda cuda_arch=90` build fails on ARM and AMD64 using:
* spack@develop (c9ec5fb9ac565033c7db643f0b51da5dc9315ac0 from `Sat Aug 19 00:46:50 2023 +0200`)
* Ubuntu 20.04, ARM64 (Also fails on AMD64)
* GCC 11.4.0
* CUDA 12.1.1
* Docker container image `ecpe4s/ubuntu20.04-runner-arm64-gcc-11.4:2023.08.01`

<details><summary>Spack environment</summary>

```
spack:
  view: false

  concretizer:
    reuse: false
    unify: false

  packages:
    all:
      require: '%gcc@11.4.0'
      providers:
        blas: [openblas]
        mpi: [mpich]
      target: [aarch64]
      variants: +mpi
    binutils:
      variants: +ld +gold +headers +libiberty ~nls
    elfutils:
      variants: +bzip2 ~nls +xz
    hdf5:
      variants: +fortran +hl +shared
    libfabric:
      variants: fabrics=sockets,tcp,udp,rxm
    libunwind:
      variants: +pic +xz
    openblas:
      variants: threads=openmp
    trilinos:
      variants: +amesos +amesos2 +anasazi +aztec +belos +boost +epetra +epetraext
        +ifpack +ifpack2 +intrepid +intrepid2 +isorropia +kokkos +ml +minitensor +muelu
        +nox +piro +phalanx +rol +rythmos +sacado +stk +shards +shylu +stokhos +stratimikos
        +teko +tempus +tpetra +trilinoscouplings +zoltan +zoltan2 +superlu-dist gotype=long_long
    xz:
      variants: +pic
    mesa:
      version: [21.3.8]
    mpi:
      require: mpich
    mpich:
      require: '@4.1.2 ~wrapperrpath ~hwloc'
    ncurses:
      require: '@6.3 +termlib'
    tbb:
      require: intel-tbb
    boost:
      version: [1.79.0]
      variants: +atomic +chrono +container +date_time +exception +filesystem +graph
        +iostreams +locale +log +math +mpi +multithreaded +program_options +random
        +regex +serialization +shared +signals +stacktrace +system +test +thread +timer
        cxxstd=17 visibility=global
    libffi:
      require: "@3.4.4"
    vtk-m:
      require: "+examples"
    cuda:
      version: [11.8.0]

  compilers:
  - compiler:
      spec: gcc@11.4.0
      paths:
        cc: /usr/bin/gcc
        cxx: /usr/bin/g++
        f77: /usr/bin/gfortran
        fc: /usr/bin/gfortran
      flags: {}
      operating_system: ubuntu20.04
      target: aarch64
      modules: []
      environment: {}
      extra_rpaths: []

  specs:
  - tasmanian +cuda cuda_arch=90
```

</details>

<details><summary>Spack environment</summary>

<pre>
 -   3yh3kmj  tasmanian@7.9%gcc@11.4.0~blas+cuda~fortran~ipo~magma+mpi~openmp~python~rocm~xsdkflags build_system=cmake build_type=Release cuda_arch=90 generator=make arch=linux-ubuntu20.04-aarch64
[+]  7hxi535      ^cmake@3.26.3%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-aarch64
[+]  ef47omc          ^curl@8.1.2%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-aarch64
[+]  4jhk4yh              ^nghttp2@1.52.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  vq4o7pj              ^openssl@3.1.2%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-aarch64
[+]  wfmbgoq                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-aarch64
[+]  wr5hfa5                  ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu20.04-aarch64
[+]  gk4kscx                      ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-aarch64
[+]  fzluouk          ^ncurses@6.3%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  kzszqob          ^zlib@1.2.13%gcc@11.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-aarch64
[+]  bb4yxiu      ^cuda@12.1.1%gcc@11.4.0~allow-unsupported-compilers~dev build_system=generic arch=linux-ubuntu20.04-aarch64
[+]  hqmyh2p          ^libxml2@2.10.3%gcc@11.4.0~python build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  dmdkgje              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-aarch64
[+]  q7galae              ^xz@5.4.1%gcc@11.4.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-aarch64
[+]  whozcr3      ^gmake@4.4.1%gcc@11.4.0~guile build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  4pfsqtg          ^gnuconfig@2022-09-17%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-aarch64
[+]  ukby5wp      ^mpich@4.1.2%gcc@11.4.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-aarch64
[+]  36ha6pi          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-aarch64
[+]  4cqdqqa          ^libfabric@1.18.1%gcc@11.4.0~debug~kdreg build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-aarch64
[+]  74eccrr          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  lunrww6              ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  cybqwl7              ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  iv32a2d          ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  g7gfedb          ^yaksa@0.2%gcc@11.4.0~cuda~rocm build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  y4hg47q              ^autoconf@2.69%gcc@11.4.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-aarch64
[+]  yvf5miz              ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  zojha6y              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-aarch64
[+]  esdggm2                  ^diffutils@3.9%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  mtu7bgv                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  qcg7zcm              ^python@3.10.12%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-ubuntu20.04-aarch64
[+]  ja4d3qc                  ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-aarch64
[+]  wmuhzov                  ^expat@2.5.0%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  ormxkmq                      ^libbsd@0.11.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  sbf43az                          ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  jm67j6t                  ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  b4ymrvu                  ^gettext@0.21.1%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  h7rmakm                      ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu20.04-aarch64
[+]  zkgaasc                          ^pigz@2.7%gcc@11.4.0 build_system=makefile arch=linux-ubuntu20.04-aarch64
[+]  iojwerc                          ^zstd@1.5.5%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-aarch64
[+]  kgueuga                  ^libffi@3.4.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  evoeirh                  ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu20.04-aarch64
[+]  iptq6b7                  ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-aarch64
[+]  loeyfnm                  ^sqlite@3.42.0%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu20.04-aarch64
[+]  xqflduv                  ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-aarch64
</pre>

</details>

Error:
```
$> spack -e . install
...
==> Installing tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr [16/16]
==> No binary for tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/de/decba62e6bbccf1bc26c6e773a8d4fd51d7f3e3e534ddd386ec41300694ce5cc.tar.gz
==> No patches needed for tasmanian
==> tasmanian: Executing phase: 'cmake'
==> tasmanian: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
...
     110    [ 25%] Building CUDA object SparseGrids/CMakeFiles/Tasmanian_libsparsegrid.dir/tsgCudaKernels.cu.o
     111    cd /tmp/root/spack-stage/spack-stage-tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr/spack-build-3yh3kmj/SparseGrids && /spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder
            __/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placehold
            e/linux-ubuntu20.04-aarch64/gcc-11.4.0/cuda-12.1.1-bb4yxiu7n2jtti7pn2bmbxd7bh5rcjs6/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -DTasmanian_libsparsegrid_EXPORTS --opti
            ons-file CMakeFiles/Tasmanian_libsparsegrid.dir/includes_CUDA.rsp -O3 -DNDEBUG -std=c++11 -Xcompiler=-fPIC -MD -MT SparseGrids/CMakeFiles/Tasmanian_libsparsegrid.dir/tsgCudaKernels.cu.o -
            MF CMakeFiles/Tasmanian_libsparsegrid.dir/tsgCudaKernels.cu.o.d -x cu -c /tmp/root/spack-stage/spack-stage-tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr/spack-src/SparseGrids/tsgCudaKern
            els.cu -o CMakeFiles/Tasmanian_libsparsegrid.dir/tsgCudaKernels.cu.o
  >> 112    /tmp/root/spack-stage/spack-stage-tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr/spack-src/InterfaceTPL/tsgCudaWrappers.cpp(301): error: identifier "CUSPARSE_MV_ALG_DEFAULT" is undefined
     113          cucheck( cusparseSpMV_bufferSize(handle, CUSPARSE_OPERATION_NON_TRANSPOSE, &alpha, matdesc, xdesc, &beta, ydesc, cuda_type, CUSPARSE_MV_ALG_DEFAULT, &bsize),
     114                                                                                                                                      ^
     115
     116    1 error detected in the compilation of "/tmp/root/spack-stage/spack-stage-tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr/spack-src/InterfaceTPL/tsgCudaWrappers.cpp".
  >> 117    make[2]: *** [SparseGrids/CMakeFiles/Tasmanian_libsparsegrid.dir/build.make:318: SparseGrids/CMakeFiles/Tasmanian_libsparsegrid.dir/__/InterfaceTPL/tsgCudaWrappers.cpp.o] Error 2
     118    make[2]: *** Waiting for unfinished jobs....
     119    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr/spack-build-3yh3kmj'
  >> 120    make[1]: *** [CMakeFiles/Makefile2:376: SparseGrids/CMakeFiles/Tasmanian_libsparsegrid.dir/all] Error 2
     121    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-tasmanian-7.9-3yh3kmjds6m7uwbcuf5enqkigqkxqndr/spack-build-3yh3kmj'
  >> 122    make: *** [Makefile:149: all] Error 2
```

### Error message

<details><summary>Error message</summary>

<pre>
... see  above
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (a38535fa8ee17c6fcafeb7a4ab7665a27d2b569e)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-neoverse_n1
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/12386598/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/12386599/spack-build-out.txt)

@mkstoyanov 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/tasmanian/package.py
+++ b/var/spack/repos/builtin/packages/tasmanian/package.py
@@ -108,6 +108,9 @@ class Tasmanian(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("magma@2.4.0:", when="+magma @6.0:", type=("build", "run"))
     depends_on("magma@2.5.0:", when="+magma @7.0:", type=("build", "run"))
 
+    # https://github.com/spack/spack/issues/39536#issuecomment-1685161942
+    conflicts("^cuda@12", when="@:7.9 +cuda")
+
     conflicts("+magma", when="~cuda~rocm")  # currently MAGMA only works with CUDA
     conflicts("+cuda", when="+rocm")  # can pick CUDA or ROCm, not both
 

